### PR TITLE
Users can edit their own comments on a forum

### PR DESCRIPTION
**Pull request by Eshah7**
## Trello ticket link

[As a user, I can edit my comments on a forum.](https://trello.com/c/6cvQCGZh/39-as-a-user-i-can-edit-my-comments-on-a-forum)

## Implementation description
* Users can click into a forum and edit their own comments
* Users can choose not to continue editing their comment
* If a user edits their comment, the timestamp will be updated as well 

## Steps to test
1. Please go to http://localhost:3000/forums
2. Please select a forum where you have already commented or create a comment
3. Edit your comment
4. Wait for the page to automatically refresh and ensure that a toast message appears and your comment is edited 
5. Please ensure that the timestamp for the comment has been updated as well 

## What should reviewers focus on?
* You should only be able to edit your own comments 
* You should be able to edit a comment as many times as you would like to
* The timestamp updates whenever a comment is updated

### CODE REVIEW

**Comment by Eshah7** [comment]
[As a user, I can edit my comments on a forum.](https://trello.com/c/6cvQCGZh/39-as-a-user-i-can-edit-my-comments-on-a-forum)

**Comment by Eshah7** [comment]
[As a user, I can edit my comments on a forum.](https://trello.com/c/6cvQCGZh/39-as-a-user-i-can-edit-my-comments-on-a-forum)

**Review by c45choi — APPROVED** [review]
looks good to me, approved bro

File: src/routes/forumComments/editForumComment [hunk 1]
```diff
@@ -0,0 +1,32 @@
+const express = require('express');
```
Comment (safchow) [inline src/routes/forumComments/editForumComment]: this needs to be a .js file

File: src/routes/forumComments/editForumComment [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+const router = express.Router();
+const mysql = require('mysql');
+const config = require('../../../config.js');
+
+//this edits an exising forum's information given the forum ID
+router.post('/api/editForumComment', (req, res) => {
```
Comment (safchow) [inline src/routes/forumComments/editForumComment]: this method should be a `PUT` method instead of a `POST` method, you also need to change all calls to this API to use the `PUT` method instead, but everything else will remain the same

File: src/routes/forumComments/editForumComment [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	let connection = mysql.createConnection(config);
+
+	let sql = `UPDATE forumComments SET
```
Comment (safchow) [inline src/routes/forumComments/editForumComment]: fix formatting

File: src/components/forums/Forum.js [hunk 4]
```diff
@@ -190,6 +223,60 @@ const Forum = () => {
...
+            setCommentError(true);
+            setCommentErrorText('Please enter your comment');
+            return false;
+        } else {
+            //console.log("about to call api edit forum");
+            callApiEditComment(selectedComment);
+            setShowEditAlertMessage(true);
+            setTimeout(() => {
+                window.location.reload();
+            }, 100);
```
Comment (safchow) [inline src/components/forums/Forum.js]: this refresh delay is too short... we usually use 3 seconds (3000 ms) instead of 0.1 seconds (100 ms)

File: src/components/forums/Forum.js [hunk 5]
```diff
@@ -190,6 +223,60 @@ const Forum = () => {
...
+        }
+    }
+    const callApiEditComment = async () => {
+        const url = `${REACT_APP_API_ENDPOINT}/editForumComment`;
+        console.log(url);
+
+        const response = await fetch(url, {
+            method: "POST",
```
Comment (safchow) [inline src/components/forums/Forum.js]: you need to update the API to a `PUT` method. Here is where you would change the method being used.

File: src/components/forums/Forum.js [hunk 6]
```diff
@@ -262,8 +357,50 @@ const Forum = () => {
...
+            <TextField
+                id="edit-comment"
+                label="Edit Comment"
+                placeholder="Enter Comment"
+                value={comment}
+                onChange={onEnterComment}
+                error={commentError}
+                fullWidth
+            />
+            <FormHelperText>{commentErrorText}</FormHelperText>
```
Comment (safchow) [inline src/components/forums/Forum.js:408]: seems like there is an error catch to ensure an empty comment can't be inputted, but no error appears informing the user this is the case:
image
The field is red, but in our other forms, there is usually an error message below. This would be a good addition here.

File: src/components/forums/Forum.js [hunk 7]
```diff
@@ -190,6 +223,60 @@ const Forum = () => {
...
+            return false;
+        } else {
+            //console.log("about to call api edit forum");
+            callApiEditComment(selectedComment);
+            setShowEditAlertMessage(true);
+            setTimeout(() => {
+                window.location.reload();
+            }, 100);
+        }
+    }
```
Comment (safchow) [inline src/components/forums/Forum.js:241]: add an empty line after this line

File: src/components/forums/Forum.js (hunk 6 above)
Comment (safchow) [reply src/components/forums/Forum.js:408]: seems like you have a state above that is supposed to do this, but isn't working for some reason

File: src/components/forums/Forums.js [hunk 8]
```diff
@@ -125,4 +125,5 @@ const Forums = () => {
     )
 }
 
+
```
Comment (safchow) [inline src/components/forums/Forums.js:128]: why was this change added?

**Review by safchow — CHANGES_REQUESTED** [review]

File: src/components/forums/Forum.js (hunk 4 above)
Comment (Eshah7) [reply src/components/forums/Forum.js]: The delete uses 100ms though... so I wanted it to be consistent. But I will change it!

File: src/components/forums/Forum.js (hunk 5 above)
Comment (Eshah7) [reply src/components/forums/Forum.js]: Does this change anything else? I have never used a "PUT" method.

File: src/components/forums/Forum.js (hunk 6 above)
Comment (Eshah7) [reply src/components/forums/Forum.js:408]: Fixed, thanks!

File: src/components/forums/Forums.js (hunk 8 above)
Comment (Eshah7) [reply src/components/forums/Forums.js:128]: Oops, thanks!

File: src/components/forums/Forum.js (hunk 6 above)
Comment (Eshah7) [reply src/components/forums/Forum.js:408]: nvm, when I fix this, the APi breaks and I don't know why, so for now I have removed it. Thanks!

File: src/routes/forumComments/editForumComment (hunk 2 above)
Comment (Eshah7) [reply src/routes/forumComments/editForumComment]: I don't really understand the PUT method... can I stick with the post method, thanks!

File: src/components/forums/Forum.js (hunk 6 above)
Comment (Eshah7) [reply src/components/forums/Forum.js:408]: Update: I have fixed this issue, thank you!

File: src/components/forums/Forum.js [hunk 9]
```diff
@@ -262,8 +358,50 @@ const Forum = () => {
...
                 ))}
             </Box>
+            {selectedComment && (
+                <div>
+                    {/* Edit comment diaglog */}
+                    < Dialog open={isDialogOpen} onClose={handleCloseDialog} >
+                        <DialogTitle>Edit Your Comment</DialogTitle>
+                        <DialogContent>
+                            <Comment
+                                comment={comment}
```
Comment (gabbytang) [inline src/components/forums/Forum.js:374]: The edit Comment textfield is a little small. For example, if I want to change a comment to be a longer sentence, I can't see the majority of it as I type.

image

**Review by gabbytang — CHANGES_REQUESTED** [review]
Just the one small thing I commented, other than that looks great!

File: src/components/forums/Forum.js (hunk 9 above)
Comment (Eshah7) [reply src/components/forums/Forum.js:374]: Screenshot 2023-03-15 at 12 12 57 PM

It is full-width, so I am not sure... will check again though! thank you

File: src/components/forums/Forum.js (hunk 5 above)
Comment (safchow) [reply src/components/forums/Forum.js]: no, `PUT` methods operate similarly to `POST` methods in terms of taking in a body as a request

File: src/routes/forumComments/editForumComment (hunk 2 above)
Comment (safchow) [reply src/routes/forumComments/editForumComment]: no, I'd rather have this addressed in this PR

File: src/components/forums/Forum.js (hunk 4 above)
Comment (safchow) [reply src/components/forums/Forum.js]: I dont think the delete displays an alert message. The other functions that use alert messages do a 3 sec delay

File: src/routes/forumComments/editForumComment.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
+const express = require('express');
+const router = express.Router();
+const mysql = require('mysql');
+const config = require('../../../config.js');
+
+//this edits an exising forum's information given the forum ID
+router.post('/api/editForumComment', (req, res) => {
```
Comment (safchow) [inline src/routes/forumComments/editForumComment.js]: change this to a `PUT` method, refer to previous discussions for implementation

**Review by safchow — CHANGES_REQUESTED** [review]

File: src/routes/forumComments/editForumComment.js (hunk 10 above)
Comment (safchow) [reply src/routes/forumComments/editForumComment.js]: working on postman after changing to `PUT`:
image

**Review by safchow — CHANGES_REQUESTED** [review]
noticing some strange behaviour with the dates, when editing a comment the time posted also updates but it doesn't make sense chronologically. Here is an example of what I mean:
image
image

Two implementations I can think of to solve this:
1. (Prefered) Add a date-edited column to the db. Render this value separately on a comment if it exists. Here is an example:
image

This is the preferable method as if the comment is edited, other users can see that it was and exactly when.

2. (Not preferable) Don't update the date if the comment is edited. This is not preferable as comments can be edited with other users not knowing that were. 

Can discuss further if needed

**Review by safchow — APPROVED** [review]

**Review by gabbytang — APPROVED** [review]
